File: app/src/main/java/io/github/gmazzo/android/livewallpaper/weather/WallpaperService.java
package io.github.gmazzo.android.livewallpaper.weather;

import android.util.Log;
import android.view.SurfaceHolder;

import io.github.gmazzo.android.livewallpaper.weather.sky_manager.WeatherInfoManager;
import io.github.gmazzo.android.livewallpaper.weather.sky_manager.WeatherInfoManager.WeatherStateReceiver;

public class WallpaperService extends GLWallpaperService {
    WeatherInfoManager mWeatherInfo;

    public class WeatherWallpaperEngine extends GLEngine implements WeatherStateReceiver {
        private static final String TAG = "WeatherWallpaperEngine";

        public WeatherWallpaperEngine() {
            super();
        }

        public void onVisibilityChanged(boolean visible) {
            if (visible) {
                WallpaperService.this.mWeatherInfo = WeatherInfoManager.getWeatherInfo(WallpaperService.this, this);
                WallpaperService.this.mWeatherInfo.update(1000);
            } else if (WallpaperService.this.mWeatherInfo != null) {
                WallpaperService.this.mWeatherInfo.onStop();
            }
            super.onVisibilityChanged(visible);
        }

        public void onCreate(SurfaceHolder surfaceHolder) {
            super.onCreate(surfaceHolder);
        }

        public void onDestroy() {
            super.onDestroy();
        }

        public synchronized void updateWeatherState() {
            Log.i("HM", "updateWeatherState");
            WeatherType weather = SettingsUtils.getWeatherConditions(WallpaperService.this);

            this.renderSurfaceView.updateWeatherType(weather);
        }
    }

    public Engine onCreateEngine() {
        return new WeatherWallpaperEngine();
    }
}

File: app/src/main/kotlin/io/github/gmazzo/android/livewallpaper/weather/WallpaperService.kt
package io.github.gmazzo.android.livewallpaper.weather

import android.view.SurfaceHolder
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.MutableStateFlow
import javax.inject.Inject

@AndroidEntryPoint
class WallpaperService : GLWallpaperService() {

    @Inject
    internal lateinit var weatherConditions: MutableStateFlow<WeatherConditions>

    override fun onCreateEngine() = WeatherWallpaperEngine()

    inner class WeatherWallpaperEngine : GLEngine() {

        // TODO once all redering is refactored, go back to `onCreate` and `onDestroy` with `SupervisorJob`
        //  https://stackoverflow.com/a/63407811
        override fun onVisibilityChanged(visible: Boolean) {
            super.onVisibilityChanged(visible)

            if (visible) {
                renderSurfaceView!!.updateWeather(weatherConditions.value)
            }
        }

        override fun onCreate(surfaceHolder: SurfaceHolder) {
            super.onCreate(surfaceHolder)

            renderSurfaceView!!.updateWeather(weatherConditions.value)
        }

    }

}
